### PR TITLE
Facebook User Authorization

### PR DESCRIPTION
**Pull request by sagemlee**
-User can log in through facebook, using outh.
-Can access authorization both from the log in page, and the register page. 
-Removed validation for password, because a facebook user will not have one. 
-Development environment uses different keys than our testing environment.

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -17,14 +17,15 @@ def create
...
+
   def show
 
   end
 
   private
 
   def user_params
-    params.permit(:first_name, :last_name, :email, :password, :password_confirmation)
+    params.permit(:first_name, :last_name, :email, :password)
```
Comment (sagemlee) [inline app/controllers/users_controller.rb:28]: Validating password_confirmation as a params was causing part of the issue.

File: app/views/sessions/new.html.erb [hunk 2]
```diff
@@ -9,3 +9,5 @@
 <%= password_field_tag :password %>
```
Comment (sagemlee) [inline app/views/sessions/new.html.erb:10]: Added functionality both here and on the registration page.

File: app/views/users/new.html.erb [hunk 3]
```diff
@@ -14,8 +14,7 @@
 <%= label_tag :password %>
 <%= password_field_tag :password %>
 
-<%= label_tag :password_confirmation %>
-<%= password_field_tag :password_confirmation %>
-
 <%= submit_tag "Create User" %>
 <% end %>
+
+<%= link_to "Connect through Facebook", "/auth/facebook"%>
```
Comment (sagemlee) [inline app/views/users/new.html.erb:20]: Added functionality both here and on the log in page.

File: spec/features/users/new_spec.rb [hunk 4]
```diff
@@ -73,38 +71,10 @@
...
     fill_in :email, with: email
     fill_in :password, with: password
-    fill_in :password_confirmation, with: password
 
     click_on "Create User"
 
     expect(current_path).to eq("/users/new")
     expect(page).to have_content("First name can't be blank")
   end
-
```
Comment (sagemlee) [inline spec/features/users/new_spec.rb:83]: Removed all of the testing for confirmation of password.

**Comment by sagemlee** [comment]
We should also do some testing once this is up in heroku. Facebook gives us separate tokens for testing that respond to a "localhost" callback, and for deployment that should respond to our actual url. Just want to double check that it does what I think it will do.

File: app/controllers/sessions_controller.rb [hunk 5]
```diff
@@ -4,6 +4,13 @@ def new
   end
 
   def create
+    if request.env['omniauth.auth']
+      user = User.create_with_omniauth(request.env['omniauth.auth'])
+      session[:user_id] = user.id
+      user.save!
+      flash[:notice] = "You are now logged in as #{user.first_name}"
+      redirect_to "/profile" 
+    else
```
Comment (jpc20) [inline app/controllers/sessions_controller.rb:13]: Nice job getting this working Sage!
Let's definitely remember to talk to Dione about any advice she on the password confirmation on Monday

File: app/models/user.rb [hunk 6]
```diff
@@ -1,11 +1,23 @@
...
     obituaries.include?(obituary)
   end
+
+  def self.create_with_omniauth(auth)
+    where(email: auth.info.email).first_or_initialize do |user|
+      user.fbid = auth.uid
+      user.first_name = auth.info.first_name 
+      user.last_name = auth.info.last_name 
+      user.email = auth.info.email 
+      user.token = auth.credentials.token
```
Comment (jpc20) [inline app/models/user.rb:19]: Awesome! Maybe as an extension, we can save their profile picture or do something else with Facebook.

Comment (sagemlee) [reply app/models/user.rb:19]: Stella similarly suggested doing something more with facebook functionality. Thats the only real reason I stored the facebook id.

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (stellakunzang) [reply app/controllers/users_controller.rb:28]: glad we were able to come up with a simple solve for this! I agree with Jack that asking Dione on Monday is a good idea to check and make sure this is ok in terms of security.

File: app/models/user.rb (hunk 6 above)
Comment (stellakunzang) [reply app/models/user.rb:19]: I think in a distant extension far down the road connecting with facebook friends on the app in order to be updated about deaths without having to be on facebook would be great, as well as importing obituaries and memorial posts/tributes. I'm really just looking for a way to get off facebook completely and not have to worry about missing this specific kind of news.

File: app/views/sessions/new.html.erb [hunk 7]
```diff
@@ -9,3 +9,5 @@
...
 
 <%= submit_tag "Submit" %>
+
+<%= link_to "Connect through Facebook", "/auth/facebook"%>
```
Comment (stellakunzang) [inline app/views/sessions/new.html.erb:13]: Yay! We are going to make this button look so good :)

File: app/controllers/sessions_controller.rb (hunk 5 above)
Comment (EricLarson2020) [reply app/controllers/sessions_controller.rb:13]: Great work here sage!

File: app/controllers/users_controller.rb (hunk 1 above)
Comment (EricLarson2020) [reply app/controllers/users_controller.rb:28]: indeed, nice work Sage

**Pull request merged**